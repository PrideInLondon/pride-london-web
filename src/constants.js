import { colors } from './theme/colors'

const constants = {
  accessibilityOptions: [
    'Step free access',
    'Wheelchair accessibile',
    'Accessible Toilets',
    'BSL Interpreter',
    'Hearing loop installed',
    'Complimentary ticket for personal assistant',
    'Service animals welcome',
    'Accessible parking spaces at venue',
  ],
  eventCategories: [
    'Cabaret and Variety',
    'Talks and Debates',
    'Community',
    'Film and Screenings',
    'Plays and Theatre',
    'Social and Networking',
    'Nightlife',
    'Exhibition and Tours',
    'Sports and Activities',
    'Health',
    'Music',
  ],
  venueDetails: ['Gender neutral toilets', 'Outdoors', 'Indoors'],
  area: ['Central', 'North', 'East', 'South', 'West'],
  audience: ['Youth', '16+', '18+', 'Families'],
  timeOfDay: ['Morning', 'Afternoon', 'Evening'],
  sponsorLevels: {
    headline: 'Headline',
    gold: 'Gold',
    silver: 'Silver',
    bronze: 'Bronze',
    partners: 'Partners',
    supporters: 'Supporters',
    mediaPartners: 'Media Partners',
    digitalPartners: 'Digital Partners',
    unlisted: 'Unlisted',
  },
  itemsToLoad: 24,
  dateFormat: 'DD/MM/YYYY',
  articleCategories: [
    {
      title: 'All Articles',
      hexColour: colors.yellow,
    },
    {
      title: 'Events',
      hexColour: colors.tomato,
    },
    {
      title: 'Fundraising',
      hexColour: colors.pink,
    },
    {
      title: 'General',
      hexColour: colors.greyBlue,
    },
    {
      title: 'Partnerships',
      hexColour: colors.bondiBlue,
    },
    {
      title: 'Research',
      hexColour: colors.eucalyptusGreen,
    },
    {
      title: 'Views',
      hexColour: colors.lemonGreen,
    },
  ],
  paradeGroupCategories: [
    {
      title: 'All groups',
      api: 'MAGIC_VALUE_NOT_USED',
      hexColour: colors.yellow,
    },
    {
      title: 'Arts',
      api: 'ARTS',
      hexColour: colors.tomato,
    },
    {
      title: 'Business',
      api: 'BUSINESS',
      hexColour: colors.pink,
    },
    {
      title: 'Charity',
      api: 'CHARITIES',
      hexColour: colors.greyBlue,
    },
    {
      title: 'Community group',
      api: 'COMMUNITY_GROUP',
      hexColour: colors.bondiBlue,
    },
    {
      title: 'Politics',
      api: 'POLITICS',
      hexColour: colors.darkPurple,
    },
    {
      title: 'Profession',
      api: 'PROFESSIONS',
      hexColour: colors.eucalyptusGreen,
    },
    {
      title: 'Sports',
      api: 'SPORTS',
      hexColour: colors.fuscia,
    },
    {
      title: 'Other',
      api: 'OTHER',
      hexColour: colors.skyBlue,
    },
  ],
  filterTypes: ['checkbox', 'radio'],
  communityPartnerCategories: [
    {
      title: 'Everything',
      hexColour: colors.yellow,
    },
    {
      title: 'Drink',
      hexColour: colors.tomato,
    },
    {
      title: 'Eat',
      hexColour: colors.pink,
    },
    {
      title: 'Party',
      hexColour: colors.greyBlue,
    },
    {
      title: 'Shop',
      hexColour: colors.eucalyptusGreen,
    },
    {
      title: 'Sleep',
      hexColour: colors.fuscia,
    },
  ],
}

export default constants
