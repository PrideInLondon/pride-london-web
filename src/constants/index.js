const theme = require('../theme/theme')

const constants = {
  accessibilityOptions: [
    'Step free access',
    'Wheelchair accessibile',
    'Accessible Toilets',
    'BSL Interpreter',
    'Hearing loop installed',
    'Complimentary ticket for personal assistant',
    'Service animals welcome',
    'Accessible parking spaces at venue',
  ],
  eventCategories: [
    'Cabaret and Variety',
    'Talks and Debates',
    'Community',
    'Film and Screenings',
    'Plays and Theatre',
    'Social and Networking',
    'Nightlife',
    'Exhibition and Tours',
    'Sports and Activities',
    'Health',
    'Music',
  ],
  venueDetails: ['Gender neutral toilets', 'Outdoors', 'Indoors'],
  area: ['Central', 'North', 'East', 'South', 'West'],
  audience: ['Youth', '16+', '18+', 'Families'],
  timeOfDay: ['Morning', 'Afternoon', 'Evening'],
  sponsorLevels: {
    headline: 'Headline',
    gold: 'Gold',
    silver: 'Silver',
    bronze: 'Bronze',
    partners: 'Partners',
    supporters: 'Supporters',
    mediaPartners: 'Media Partners',
    digitalPartners: 'Digital Partners',
  },
  itemsToLoad: 24,
  dateFormat: 'DD/MM/YYYY',
  articleCategories: [
    {
      title: 'All Articles',
      hexColour: theme.colors.yellow,
    },
    {
      title: 'Events',
      hexColour: theme.colors.tomato,
    },
    {
      title: 'Fundraising',
      hexColour: theme.colors.pink,
    },
    {
      title: 'General',
      hexColour: theme.colors.greyBlue,
    },
    {
      title: 'Partnerships',
      hexColour: theme.colors.bondiBlue,
    },
    {
      title: 'Research',
      hexColour: theme.colors.eucalyptusGreen,
    },
    {
      title: 'Views',
      hexColour: theme.colors.lemonGreen,
    },
  ],
  paradeGroupCategories: [
    {
      title: 'All groups',
      api: 'MAGIC_VALUE_NOT_USED',
      hexColour: theme.colors.yellow,
    },
    {
      title: 'Arts',
      api: 'ARTS',
      hexColour: theme.colors.tomato,
    },
    {
      title: 'Business',
      api: 'BUSINESS',
      hexColour: theme.colors.pink,
    },
    {
      title: 'Charity',
      api: 'CHARITIES',
      hexColour: theme.colors.greyBlue,
    },
    {
      title: 'Community group',
      api: 'COMMUNITY_GROUP',
      hexColour: theme.colors.bondiBlue,
    },
    {
      title: 'Politics',
      api: 'POLITICS',
      hexColour: theme.colors.darkPurple,
    },
    {
      title: 'Profession',
      api: 'PROFESSIONS',
      hexColour: theme.colors.eucalyptusGreen,
    },
    {
      title: 'Sports',
      api: 'SPORTS',
      hexColour: theme.colors.fuscia,
    },
    {
      title: 'Other',
      api: 'OTHER',
      hexColour: theme.colors.skyBlue,
    },
  ],
  filterTypes: ['checkbox', 'radio'],
  communityPartnerCategories: [
    {
      title: 'Everything',
      hexColour: theme.colors.yellow,
    },
    {
      title: 'Drink',
      hexColour: theme.colors.tomato,
    },
    {
      title: 'Eat',
      hexColour: theme.colors.pink,
    },
    {
      title: 'Party',
      hexColour: theme.colors.greyBlue,
    },
    {
      title: 'Shop',
      hexColour: theme.colors.eucalyptusGreen,
    },
    {
      title: 'Sleep',
      hexColour: theme.colors.fuscia,
    },
  ],
}

module.exports = constants
