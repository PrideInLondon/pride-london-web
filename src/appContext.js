import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import { filterPastEvents, getDuration, sanitizeDates } from './events/helpers'
import constants from './constants'

const AppContext = React.createContext()
const { Consumer } = AppContext

const initialState = {
  events: [],
}

class Provider extends Component {
  constructor() {
    super()
    this.state = {
      ...initialState,
    }
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.events === prevState.events) return null

    const allEventOccurences = []

    nextProps.events.forEach(event => {
      if (!event.node.recurrenceDates) {
        allEventOccurences.push(event)
      } else {
        const recurrenceDates = sanitizeDates([
          moment(event.node.startTime).format(constants.dateFormat),
          ...event.node.recurrenceDates,
        ])
        const time = moment(event.node.startTime).format('HH:mm')
        const duration = getDuration(event.node.startTime, event.node.endTime)

        recurrenceDates.forEach(date => {
          const copy = JSON.parse(JSON.stringify(event))

          copy.node.startTime = moment(
            `${date} ${time}`,
            'DD/MM/YYYY hh:mm'
          ).format()
          copy.node.endTime = moment(copy.node.startTime)
            .add(duration, 'milliseconds')
            .format()

          allEventOccurences.push(copy)
        })
      }
    })
    return {
      events: allEventOccurences
        .filter(filterPastEvents)
        .sort((a, b) =>
          a.node.startTime < b.node.startTime
            ? -1
            : a.node.startTime > b.node.startTime
            ? 1
            : 0
        ),
    }
  }

  render() {
    return (
      <AppContext.Provider
        value={{
          state: this.state,
        }}
      >
        {this.props.children}
      </AppContext.Provider>
    )
  }
}

Provider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  events: PropTypes.array, // eslint-disable-line
}

Provider.defaultProps = {
  events: [],
}

export { AppContext, Provider, Consumer }
