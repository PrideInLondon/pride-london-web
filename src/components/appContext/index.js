import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import {
  filterByDate,
  filterByFree,
  filterByCategory,
  filterByArea,
  filterByTime,
  filterPastEvents,
  getDuration,
  sanitizeDates,
} from '../../features/events/helpers/index'
import constants from '../../constants'

const AppContext = React.createContext()
const { Consumer } = AppContext

function getInitialFilterState() {
  return {
    startDate: null,
    endDate: null,
    free: false,
    eventCategories: [],
    venueDetails: [],
    audience: [],
    accessibilityOptions: [],
    area: [],
    timeOfDay: [],
  }
}

const initialState = {
  events: [],
  filterOpen: null,
  eventsToShow: constants.itemsToLoad,
  filters: getInitialFilterState(),
}

class Provider extends Component {
  constructor() {
    super()
    this.state = {
      ...initialState,
    }
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.events !== prevState.events) {
      const allEventOccurences = []

      nextProps.events.forEach(event => {
        if (!event.node.recurrenceDates) {
          allEventOccurences.push(event)
        } else {
          const recurrenceDates = sanitizeDates([
            moment(event.node.startTime).format(constants.dateFormat),
            ...event.node.recurrenceDates,
          ])
          const time = moment(event.node.startTime).format('HH:mm')
          const duration = getDuration(event.node.startTime, event.node.endTime)

          recurrenceDates.forEach(date => {
            const copy = JSON.parse(JSON.stringify(event))

            copy.node.startTime = moment(
              `${date} ${time}`,
              'DD/MM/YYYY hh:mm'
            ).format()
            copy.node.endTime = moment(copy.node.startTime)
              .add(duration, 'milliseconds')
              .format()

            allEventOccurences.push(copy)
          })
        }
      })
      return {
        events: allEventOccurences
          .filter(filterPastEvents)
          .sort((a, b) =>
            a.node.startTime < b.node.startTime
              ? -1
              : a.node.startTime > b.node.startTime
              ? 1
              : 0
          ),
      }
    }
  }

  getDatepickerValues = ({ startDate, endDate }) => {
    this.setState(prevState => ({
      ...prevState,
      filters: {
        ...prevState.filters,
        startDate,
        endDate,
      },
    }))
  }

  setDate = (dateToSet, dateToGet) => {
    this.setState(prevState => ({
      ...prevState,
      filters: {
        ...prevState.filters,
        [dateToSet]: prevState.filters[dateToGet],
      },
    }))
  }

  getCheckboxBool = (name, checked) => {
    this.setState(prevState => ({
      ...prevState,
      filters: {
        ...prevState.filters,
        free: checked,
      },
    }))
  }

  getCheckboxSetValues = (e, name) => {
    const state = {
      ...this.state,
      filters: { ...this.state.filters },
    }

    if (
      e.target.checked &&
      state.filters[name].indexOf(e.target.value) === -1
    ) {
      state.filters[name].push(e.target.value)
    } else {
      const index = state.filters[name].indexOf(e.target.value)
      if (index > -1) {
        this.state.filters[name].splice(index, 1)
      }
    }

    this.setState(state)
  }

  clearFilters = () => {
    this.setState({
      filterOpen: null,
      filters: getInitialFilterState(),
    })
  }

  closeSiblingFilters = (filterName, isOpen) => {
    if (isOpen && filterName != this.state.filterOpen) {
      this.setState(prevState => ({
        ...prevState,
        filterOpen: filterName,
      }))
    } else {
      this.setState(prevState => ({
        ...prevState,
        filterOpen: null,
      }))
    }
  }

  filterEvents = () => {
    const filteredEvents = this.state.events
      .filter(filterByDate, {
        startDate: this.state.filters.startDate,
        endDate: this.state.filters.endDate,
      })
      .filter(filterByFree, this.state.filters.free)
      .filter(filterByCategory, {
        array: this.state.filters.eventCategories,
        key: 'eventCategories',
      })
      .filter(filterByCategory, {
        array: this.state.filters.venueDetails,
        key: 'venueDetails',
      })
      .filter(filterByCategory, {
        array: this.state.filters.accessibilityOptions,
        key: 'accessibilityOptions',
      })
      .filter(filterByCategory, {
        array: this.state.filters.audience,
        key: 'audience',
      })
      .filter(filterByArea, this.state.filters.area)
      .filter(filterByTime, this.state.filters.timeOfDay)

    return filteredEvents
  }

  showMore = filteredCount => {
    if (this.state.eventsToShow < filteredCount) {
      this.setState({
        eventsToShow: this.state.eventsToShow + constants.itemsToLoad,
      })
    }
  }

  render() {
    return (
      <AppContext.Provider
        value={{
          state: this.state,
          filteredEvents: this.filterEvents(),
          actions: {
            getCheckboxBool: this.getCheckboxBool,
            getDatepickerValues: this.getDatepickerValues,
            getCheckboxSetValues: this.getCheckboxSetValues,
            clearFilters: this.clearFilters,
            closeSiblingFilters: this.closeSiblingFilters,
            showMore: this.showMore,
            setDate: this.setDate,
          },
        }}
      >
        {this.props.children}
      </AppContext.Provider>
    )
  }
}

Provider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  events: PropTypes.array, // eslint-disable-line
}

Provider.defaultProps = {
  events: [],
}

export { Provider, Consumer }
