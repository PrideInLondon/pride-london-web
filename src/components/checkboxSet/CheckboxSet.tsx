import React, { useContext } from 'react'
import { AppContext } from '../appContext'
import Checkbox from '../checkbox'
import constants from '../../constants'
import { List, ListItem } from './CheckboxSet.styles'
import { CheckboxSetProps } from './CheckboxSet.types'

const CheckboxSet = ({ filterName, sort = null }: CheckboxSetProps) => {
  const { state, actions } = useContext(AppContext)

  // @ts-ignore - refactoring of constants required
  const options = constants[filterName] || []

  const makeId = (inputId: string) => {
    const id = inputId
      .replace(/[^a-zA-Z0-9 ]/g, '')
      .replace(/ /g, '_')
      .toLowerCase()
    return `${filterName}_${id}`
  }

  if (sort) {
    options.sort()
    if (sort === 'DESC') {
      options.reverse()
    }
  }

  return (
    <List>
      {options.map((option: string) => (
        <ListItem key={makeId(option)}>
          <Checkbox
            //  @ts-ignore - refactoring of appContext required
            checked={state.filters[filterName].indexOf(option) >= 0}
            label={option}
            value={option}
            id={makeId(option)}
            name={makeId(option)}
            handleChange={e => {
              //  @ts-ignore - refactoring of appContext required
              return actions.getCheckboxSetValues(e, filterName)
            }}
          />
        </ListItem>
      ))}
    </List>
  )
}

export default CheckboxSet
