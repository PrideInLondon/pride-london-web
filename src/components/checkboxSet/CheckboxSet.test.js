import React from 'react'
import { mount } from 'enzyme'
import 'jest-styled-components'
import constants from '../../constants'
import Checkbox from '../checkbox'
import { Provider } from '../appContext'
import CheckboxSet from './CheckboxSet'

// SVGs need to be mocked as the SVG output causes Babel to throw
jest.mock('../../theme/assets/images/icon-check.svg', () => 'foo')

describe('CheckboxSet', () => {
  const filterName = 'eventCategories'
  const options = constants[filterName] || []

  it('should render list properly', () => {
    const wrapper = mount(
      <Provider>
        <CheckboxSet filterName={filterName} sort="DESC" />
      </Provider>
    )
    expect(wrapper.find(Checkbox)).toHaveLength(options.length)
  })

  it('should check filter properly', () => {
    const wrapper = mount(
      <Provider>
        <CheckboxSet filterName={filterName} sort="DESC" />
      </Provider>
    )
    const checkbox = wrapper.find(Checkbox).first()
    const input = checkbox.find('input[type="checkbox"]')
    expect(input.props().checked).toEqual(false)
    checkbox.simulate('click')
    setTimeout(() => {
      expect(input.props().checked).toEqual(true)
    })
  })
})
