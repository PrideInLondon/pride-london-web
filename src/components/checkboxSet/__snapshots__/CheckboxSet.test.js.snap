// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxSet renders 1`] = `<CheckboxSetstyles__List />`;
